### ISSUE TITLE
Feature Request: Expose some general/resource stats from Realm core.

### ISSUE THREAD

**Issue by diegomontoya** [realm/realm-java#3181]
What:

Expose global Resource usage summary from the realm core (which is currently a black box) at the point of the call. For example:

1) How many files/fds are currently opened/held by realm core. 
2) How many active mmap() are there and the size of each call.
2) How many realms are currently open.  For each realm, how many different versions are currently open but not closed due to each thread/transaction holding on a different version?
3) How many realm queries are active (not released)
4) Any other stats info that may be relevant to the dev to understand resource usage of the non-java portion of the library. 

The above are some things just came to to the top of my head. It may or may not be possible but any info would be great. Think of this as a overview resource usage point-in-time for Realm core. 

Why:

A) General stats. regardless how minimal,  can really reveal an implementation bug by the end user. Realms not closing correctly, queries held and not released, too many different threads are opening realms. 

B) This would also help developer create better android tests when we know at a specific point, barring GC and finalizer lag, that there should be X realms open, X for other stats. It help devs create better tests for code using realm library by comparing current values to expected internal states. 

C) It is sometimes extremely difficult to debug bugs caused by resource usage caused by  a .so library: too much ram used, too many fds, etc, if the library does not expose some stats info.  Android and especially version <= 4.X can behave wildly under extremely resource exhaustion and crash at points that has no relevancy due to the core cause. If we can log the crashes with the the realm core internal stats, it would help give devs get  a closer version of the 100% view of the issue at hand and not just where it crashed, which sometimes is really useless. 

Some the above is possible by extending Realm or wrapping realm opens/closes transaction/open/close/do our own ref/stat counting but that's a lot of abstraction when the core has already all the info plus other info we can't possibly get currently. 

Sorry for the verbosity but hopefully the above gives a general idea of what I propose and reasoning behind it.

**Comment by beeender** [comment]
I totally agree. Although most of data you asked can be recorded in java layer, it is still good to have the feature implemented in our underlying engine. I will create an issue in our storage engine repo.

**Comment by beeender** [comment]
> B) This would also help developer create better android tests when we know at a specific point, barring GC and finalizer lag, that there should be X realms open, X for other stats. It help devs create better tests for code using realm library by comparing current values to expected internal states. 

BTW, Finalizer will be killed soon by #3144

**Comment by diegomontoya** [comment]
@beeender Awesome! Very much looking forward to #3144. 

>  Although most of data you asked can be recorded in java layer, it is still good to have the feature implemented in our underlying engine. 

The stats can be added/recorded by the java layer but the direct core stats would be best since it is not dependent on any wrapper code which can introduce bugs themselves by losing objects that hold the native refs. Might even help unit testing the library itself as well post #3144.

**Comment by nateridderman** [comment]
This would be super useful for debugging an issue where Realm's db file keeps increasing in size when it's running in the background for a long time. For instance, now that core can report on aggregate table size, can that be exposed as part of this? https://github.com/realm/realm-core/pull/2591.

**Comment by beeender** [comment]
@nateridderman Maybe https://github.com/realm/realm-java/pull/4857 would help in your case?
